src/shell/shell_batch.cpp

```diff
@@ -19,275 +19,211 @@
-		DOS_CloseFile(file_handle);
-		return false; // Parent deletes this BatchFile on negative return
-	}
-	if (!strlen(temp)) goto emptyline;
-	if (temp[0]==':') goto emptyline;
-
-	// Lambda that copies the src to cmd_write, provided the result fits
-	// within CMD_MAXLINE while taking the existing line into consideration.
-	// Finally it moves the cmd_write pointer ahead by the length copied.
-	auto append_cmd_write = [&cmd_write, &line](const char *src) {
-		const auto src_len = strlen(src);
-		const auto req_len = cmd_write - line + static_cast<int>(src_len) + 1;
-		if (src_len && req_len < CMD_MAXLINE) {
-			safe_strncpy(cmd_write, src, req_len);
-			cmd_write += src_len;
+
+		// EOF
+		if (bytes_read == 0) {
+			break;
 		}
-	};
-
-	/* Now parse the line read from the bat file for % stuff */
-	cmd_write=line;
-	char * cmd_read=temp;
-	while (*cmd_read) {
-		if (*cmd_read == '%') {
-			cmd_read++;
-			if (cmd_read[0] == '%') {
-				cmd_read++;
-				if (((cmd_write - line) + 1) < (CMD_MAXLINE - 1))
-					*cmd_write++ = '%';
-				continue;
-			}
-			if (cmd_read[0] == '0') {  /* Handle %0 */
-				const char *file_name = cmd->GetFileName();
-				cmd_read++;
-				append_cmd_write(file_name);
-				continue;
-			}
-			char next = cmd_read[0];
-			if(next > '0' && next <= '9') {
-				/* Handle %1 %2 .. %9 */
-				cmd_read++; //Progress reader
-				next -= '0';
-				if (cmd->GetCount()<(unsigned int)next) continue;
-				std::string word;
-#if (CHAR_MIN < 0) // char is signed
-				assert(next >= 0);
-#endif
-				if (!cmd->FindCommand(static_cast<unsigned>(next), word))
-					continue;
-				append_cmd_write(word.c_str());
-				continue;
-			} else {
-				/* Not a command line number has to be an environment */
-				char * first = strchr(cmd_read,'%');
-				/* No env after all. Ignore a single % */
-				if (!first) {
-					/* *cmd_write++ = '%';*/
-					//check if input contains cycles + max/auto  and that next character is space or empty
-					//If so, don't ignore it. This way cycles can still be set from within batch files
-					char peak = *(cmd_read);
-					if (peak == 0 || peak == ' ' || peak == '\r' || peak == '\n') {
-						strncpy(temp_cycles_hack,temp,cmd_read-temp);
-						temp_cycles_hack[cmd_read-temp] = 0;
-						upcase(temp_cycles_hack);
-						const char* cycles_test_cycles = strstr(temp_cycles_hack,"CYCLES");
-						if (cycles_test_cycles) {
-							const char* cycles_test_max = strstr(cycles_test_cycles,"MAX");
-							const char* cycles_test_auto = strstr(cycles_test_cycles,"AUTO");
-							if ( cycles_test_max  || cycles_test_auto )	{
-								   if (((cmd_write - line) + 1) < (CMD_MAXLINE - 1))
-									   *cmd_write++ = '%';
-							}
-						}
-					}
-					continue;
-				}
-				*first++ = 0;
-				std::string env;
-				if (shell->GetEnvStr(cmd_read,env)) {
-					const char* equals = strchr(env.c_str(),'=');
-					if (!equals) continue;
-					equals++;
-					append_cmd_write(equals);
-				}
-				cmd_read = first;
-			}
+
```

Comment:
This function was doing a lot! So much easier to read and surely safer (as you mentioned the previous code could crash). :partying_face: 

Do you have a a bunch of batch files that you're testing with, to be sure that we're not regression or dropping functionality?

Also, there are some shell unit tests - maybe this batch file handling could be added, and you could add some positive / negative / errors etc type tests to exercise all the corner cases?

Reply:
Fully agree it doesn't make sense to battle with gmocking the whole-program and thread a needle into these functions, , @MeAreJeenius.

The flatter more stand-alone refactored BatchFile class sounds great!